### PR TITLE
 Exclude the load balancing loss of padding tokens in Mixtral-8x7B

### LINKED ISSUE #28505: Exclude the load balancing loss of padding tokens in Mixtral-8x7B
### Feature request

The auxiliary loss in Mixtral-MoE shouldn't **include the loss from padding tokens**. 

### Motivation

I think it is better to change the function 
[load_balancing_loss_func](https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py#L77) by adding an additional parameter: `attention_mask` and change the implementation inside to remove the loss from padding tokens

### Your contribution

I would be happy to review the PR implemeting this feature !

### PR DESCRIPTION
**Pull request by khaimt**
# What does this PR do?
This PR implements excluding the load balancing loss of padding tokens in Mixtral-8x7B

Fixes #28505

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@ArthurZucker

### CODE REVIEW

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -95,21 +100,45 @@ def load_balancing_loss_func(gate_logits: torch.Tensor, num_experts: torch.Tenso
     if gate_logits is None or not isinstance(gate_logits, tuple):
         return 0
 
+    batch_size, sequence_length = attention_mask.shape
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:103]: attention mask can be None, let's make sure this is taken into account, I'd rather we don't create / compute everything with a dummy mask if no mask is provided!

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 2]
```diff
@@ -1347,10 +1376,13 @@ def forward(
...
             aux_loss = load_balancing_loss_func(
-                outputs.router_logits if return_dict else outputs[-1], self.num_experts, self.num_experts_per_tok
+                outputs.router_logits if return_dict else outputs[-1],
+                attention_mask if attention_mask is not None else torch.ones_like(input_ids),
+                self.num_experts,
+                self.num_experts_per_tok,
             )
             if labels is not None:
-                loss += self.router_aux_loss_coef * aux_loss
+                loss += self.router_aux_loss_coef * aux_loss.to(loss.device)  # make sure to reside in the same device
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:1385]: good catch

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 1 above)
Comment (khaimt) [reply src/transformers/models/mixtral/modeling_mixtral.py:103]: Actually, I already **checked if attention_mask is None** at where we use the function:
https://github.com/khaimt/transformers/blob/exclude_padding_tokens_in_aux_loss_Mixtral_Moe/src/transformers/models/mixtral/modeling_mixtral.py#L1380

Comment (ArthurZucker) [reply src/transformers/models/mixtral/modeling_mixtral.py:103]: I don't think computing more things is better than skipping this!

**Review by ArthurZucker — COMMENTED** [review]
let's rather skip the attention part if we don't have a mask no? Less compute and no changes for people who don't want extra computation

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 1 above)
Comment (khaimt) [reply src/transformers/models/mixtral/modeling_mixtral.py:103]: Yes, Let me update the code to skip computing more when attention_mask is None

Comment (khaimt) [reply src/transformers/models/mixtral/modeling_mixtral.py]: @ArthurZucker I have just updated the code for skipping computing more when attention_mask is None

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the update. Can you add a good test to make sure this behaves as expected? 🤗

**Comment by khaimt** [comment]
> Thanks for the update. Can you add a good test to make sure this behaves as expected? 🤗

Sure, I will add a test to make sure that the implementation is correct

**Comment by khaimt** [comment]
@ArthurZucker I don't know why The test is failed on github but was passed in my local test.
In my local test, I only changed the test_mixtral.py so I ran:
pytest --picked
And test was passed, but was failed on github

**Comment by congruency** [comment]
> I don't know why The test is failed on github but was passed in my local test. 

From details of [check_code_quality](https://circleci.com/gh/huggingface/transformers/1066145):
tests/models/mixtral/test_modeling_mixtral.py:19:8: F401 [*] `random` imported but unused

Unused import of the random module in the test file.

**Comment by khaimt** [comment]
> > I don't know why The test is failed on github but was passed in my local test.
> 
> From details of [check_code_quality](https://circleci.com/gh/huggingface/transformers/1066145): tests/models/mixtral/test_modeling_mixtral.py:19:8: F401 [*] `random` imported but unused
> 
> Unused import of the random module in the test file.

I mean this assert: https://app.circleci.com/pipelines/github/huggingface/transformers/82724/workflows/61975ffc-0bdb-4aa7-935e-70f69f11e748/jobs/1066148/parallel-runs/0/steps/0-117

**Comment by congruency** [comment]
Some thoughts:
- Regarding swapping the sign for `∣actual−expected∣≤atol+rtol⋅∣expected∣`.. You aren't getting the fabs() of result.aux_loss before multiplying. 
- There's also an [assertNotAlmostEqual](https://github.com/huggingface/transformers/blob/3f69f415adcbdaedec154ba8eac220ef3276975d/tests/models/encoder_decoder/test_modeling_encoder_decoder.py#L817) function for comparing if they're not close.
- Specifying the device to use might give different results on different systems

Thank you for doing this btw!

**Comment by khaimt** [comment]
> Some thoughts:
> 
>     * Regarding `∣actual−expected∣>atol+rtol⋅∣expected∣`.. You aren't getting the fabs() of result.aux_loss before multiplying.
> 
>     * There's also an [assertNotAlmostEqual](https://github.com/huggingface/transformers/blob/3f69f415adcbdaedec154ba8eac220ef3276975d/tests/models/encoder_decoder/test_modeling_encoder_decoder.py#L817) function for comparing if they're not close.
> 
>     * Specifying the device to use might give different results on different systems
> 
> 
> Thank you for doing this btw!
> 
> Edit for clarity

Hi @congruency Thank you for your comments
Actually we no need to get the fabs() of result.aux_loss, because it is already positive, and actually the value is around 2. You can see the value from previous assert: 
https://github.com/khaimt/transformers/blob/exclude_padding_tokens_in_aux_loss_Mixtral_Moe/tests/models/mixtral/test_modeling_mixtral.py#L478 
Oh, thankyou for reminding me of: assertNotAlmostEqual, I will fix my code with this suggestion

**Comment by khaimt** [comment]
@ArthurZucker The tests have been fully added and passed, can you review ?

**Review by Bazzimore — APPROVED** [review]
This is very good and I need to work with this one

**Comment by Bazzimore** [comment]
Am a business

**Comment by Bazzimore** [comment]
UC eye care center

**Comment by ArthurZucker** [comment]
Sorry for the delay will review!

File: tests/models/mixtral/test_modeling_mixtral.py [hunk 3]
```diff
@@ -476,6 +475,24 @@ def test_load_balancing_loss(self):
         self.assertEqual(result.router_logits[0].shape, (91, config.num_local_experts))
         torch.testing.assert_close(result.aux_loss.cpu(), torch.tensor(2, dtype=torch.float32), rtol=1e-2, atol=1e-2)
 
+        # First, we make sure that adding padding tokens doesn't change the loss
+        # loss(input_ids, attention_mask=None) == loss(input_ids + padding, attention_mask=attention_mask_with_padding)
+        pad_leng = 1000
```
Comment (ArthurZucker) [inline tests/models/mixtral/test_modeling_mixtral.py:480]:
```suggestion
        pad_length = 1000
```

File: tests/models/mixtral/test_modeling_mixtral.py [hunk 4]
```diff
@@ -476,6 +475,24 @@ def test_load_balancing_loss(self):
...
+        # Add padding tokens (assume that pad_token_id=1) to input_ids
+        padding_block = torch.ones(input_ids.shape[0], pad_leng, dtype=torch.int32).to(torch_device)
+        padded_input_ids = torch.cat((padding_block, input_ids), dim=1)  # this is to simulate padding to the left
+        padded_attention_mask = padded_input_ids.ne(1).to(torch_device)
+
+        padded_result = model(padded_input_ids, attention_mask=padded_attention_mask)
+        torch.testing.assert_close(result.aux_loss.cpu(), padded_result.aux_loss.cpu(), rtol=1e-2, atol=1e-2)
```
Comment (ArthurZucker) [inline tests/models/mixtral/test_modeling_mixtral.py:487]: can't we have a better threshold here? (tol)

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 5]
```diff
@@ -74,7 +74,9 @@
 _CONFIG_FOR_DOC = "MixtralConfig"
 
 
-def load_balancing_loss_func(gate_logits: torch.Tensor, num_experts: torch.Tensor = None, top_k=2) -> float:
+def load_balancing_loss_func(
+    gate_logits: torch.Tensor, attention_mask: Optional[torch.Tensor], num_experts: torch.Tensor = None, top_k=2
+) -> float:
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:79]: let's put the attention mask last here, this way we don't break previous calls!

**Review by ArthurZucker — COMMENTED** [review]
LGTM thank you for iterating

File: tests/models/mixtral/test_modeling_mixtral.py (hunk 4 above)
Comment (khaimt) [reply tests/models/mixtral/test_modeling_mixtral.py]: Yes we can, I have just reduced to 1e-4

**Comment by khaimt** [comment]
@ArthurZucker I have just fixed my PR:
+ move attention_mask to the last position
+ pad_leng --> pad_length
+ reduce the tol for assert_close

**Review by ArthurZucker — APPROVED** [review]
Thank you for adding this feature!
